Create "direct" install sub command that will be called in a posttrans on package mode installations

yasminvalim opened the pull request:
TICKET: As part of https://fedoraproject.org/wiki/Changes/BootLoaderUpdatesPhase1, for package mode installations, we will update the bootloader (copy from /usr to boot or ESP) as part of the posttrans scriptlet.

This is a slightly different path as the current update path where we have the systemd-run logic, and we notably don't have any prepared metadata.

We can start by creating a command that does a simple copy (like the scripts do) and then iterate on merging the logic between the package mode and the image mode use case.

See change and current sample/basic installation scripts:

https://fedoraproject.org/wiki/Changes/BootLoaderUpdatesPhase1
https://src.fedoraproject.org/rpms/grub2/pull-request/136
https://src.fedoraproject.org/rpms/shim/pull-request/4

See related issue: https://github.com/coreos/bootupd/issues/926

HuijingHei commented:
> [#1021 (comment)](https://github.com/coreos/bootupd/pull/1021#issuecomment-3550193095):
> 
> > @travier the command that we want to use is for package mode, also for image mode, is this right?
> 
> This command will be used exclusively on package mode.

Thank you for the confirmation, then it would be easy for us to add such tmt test.

travier commented:
I have not reviewed the code yet, but in the meantime, can you test that this change works when you run it as part of a GRUB update? See `%posttrans` in the GRUB package: https://src.fedoraproject.org/rpms/grub2/blob/rawhide/f/grub2.spec#_430

yasminvalim commented:
Hey  @HuijingHei, I can take a look on how do TMT tests and apply it here. Do you have any examples of this kind of test?

Thanks @travier for your support!

HuijingHei commented:
@yasminvalim I am doing tests with [add-post](https://github.com/coreos/bootupd/compare/main...HuijingHei:bootupd:add-post), also rebase this PR, build package for f44, and install it on package mode, failed when running posttrans, could you help to check? Thank you!
```
[root@citest-1 ~]# dnf install -y /mnt/shared/bootupd/bootupd-0.2.32.23.gd2b568e-1.fc44.x86_64.rpm
Updating and loading repositories:
Repositories loaded.
Package                                                   Arch          Version                                                    Repository                            Size
Installing:
 bootupd                                                  x86_64        0:0.2.32.23.gd2b568e-1.fc44                                @commandline                       3.5 MiB

Transaction Summary:
 Installing:         1 package

Total size of inbound packages is 1 MiB. Need to download 0 B.
After this operation, 4 MiB extra will be used (install 4 MiB, remove 0 B).
Running transaction
[1/3] Verify package files                                                                                                           100% |  76.0   B/s |   1.0   B |  00m00s
[2/3] Prepare transaction                                                                                                            100% |  16.0   B/s |   1.0   B |  00m00s
[3/3] Installing bootupd-0:0.2.32.23.gd2b568e-1.fc44.x86_64                                                                          100% |   5.4 MiB/s |   3.5 MiB |  00m01s
>>> Running %posttrans scriptlet: bootupd-0:0.2.32.23.gd2b568e-1.fc44.x86_64
>>> Finished %posttrans scriptlet: bootupd-0:0.2.32.23.gd2b568e-1.fc44.x86_64
>>> Scriptlet output:
>>> error: unrecognized subcommand 'copy-to-boot'
>>> 
>>> Usage: bootupctl backend [OPTIONS]
>>> 
>>> For more information, try '--help'.
>>> 
Warning: skipped OpenPGP checks for 1 package from repository: @commandline
Complete!
```

travier commented:
> Just one small question about this, after install the new bootupctl, `/boot/efi` is unmounted, maybe we should not unmount (the same as before installation)? WDYT? @travier

Ah yes, when we run that in a systemd service unit we can mount/unmount things as we are running in our own mount namespace but when running as part of an RPM transaction we should not touch mount points.

cgwalters commented on src/efi.rs:
```diff
@@ -183,6 +191,89 @@ impl Efi {
+
+    /// Copy from /usr/lib/efi to boot/ESP.
+    fn package_mode_copy_to_boot_impl(&self) -> Result<()> {
+        let sysroot = Path::new("/");
```
Why not require the caller to provide this? I can imagine use cases where we want this to operate on a non-mounted root like system recovery.

Among other things...it makes unit tests work.

cgwalters commented on src/efi.rs:
```diff
@@ -183,6 +191,89 @@ impl Efi {
+        let efi_comps = match get_efi_component_from_usr(sysroot_path, EFILIB)? {
+            Some(comps) if !comps.is_empty() => comps,
+            _ => {
+                log::debug!("No EFI components found in /usr/lib/efi");
```
I think there's an open question of if this should be configurable as an error or not.

I think it probably should be.

yasminvalim replied:
I am not sure if I understood. Do you mean we should make “no components” configurable (e.g. strict vs allow via a flag)? If yes, I can add a CLI option for it.

HuijingHei commented on src/efi.rs:
```diff
@@ -68,6 +68,8 @@ pub(crate) fn is_efi_booted() -> Result<bool> {
 pub(crate) struct Efi {
     mountpoint: RefCell<Option<PathBuf>>,
+    /// Track whether we mounted the ESP ourselves (true) or found it pre-mounted (false)
+    did_mount: RefCell<bool>,
```
> Hmm seems cleaner to have one RefCell i.e.
> 
> ```
> struct Mount {
>   path: PathBuf,
>   owned: bool
> }
> 
> struct Efi {
>   mountpoint: RefCell<Option<Mount>>
> ```
LGTM.

> That said even cleaner would be for us to use the new mount API internally

I think you meant for https://github.com/bootc-dev/bootc/blob/main/crates/mount/Cargo.toml which is using rust `edition = "2024"`, we should use the new mount once move to 2024 (now using `edition = "2021"`)

yasminvalim replied:
RefCell is now a single Mount { path, owned }, and we use the new mount API in mount_esp(). Could you take a look when you have a moment?

yasminvalim commented on src/bootupd.rs:
```diff
@@ -688,6 +688,23 @@ pub(crate) fn client_run_migrate_static_grub_config() -> Result<()> {
+pub(crate) fn copy_to_boot() -> Result<()> {
+    let all_components = get_components_impl(false);
+    if all_components.is_empty() {
+        println!("No components available for this platform.");
```
Should we keep it? I added a error when there are no components. WDYT?

yasminvalim commented on src/component.rs:
```diff
@@ -79,6 +79,11 @@ pub(crate) trait Component {
+
+    /// Copy from /usr/lib/efi to boot/ESP.
+    fn package_mode_copy_to_boot(&self) -> Result<()> {
+        Ok(())
```
Done. No default anymore. WDYT?

yasminvalim commented on src/efi.rs:
```diff
@@ -183,6 +191,89 @@ impl Efi {
+        }
+
+        // Sync filesystem
+        let efidir =
```
I applied this suggestions but not sure if its okay. Feel free to take a look.

yasminvalim commented on src/efi.rs:
```diff
@@ -948,7 +1042,7 @@ Boot0003* test";
             tmpd.atomic_write(
                 "etc/system-release",
-                "Red Hat Enterprise Linux CoreOS release 4
+                r"Red Hat Enterprise Linux CoreOS release 4
```
The test is just to ensure the parser will return the same name even when the file ends with a newline. It makes sense for you? If you think is not a good test scenario let me now so I can delete it :)

Thanks for the review and the comments :)

HuijingHei commented:
/gemini review

HuijingHei commented on src/efi.rs:
```diff
@@ -108,8 +145,16 @@ impl Efi {
             if !mnt.exists() {
                 continue;
             }
+            #[cfg(target_os = "linux")]
```
I do not remember we support target_os is not linux, maybe we should remove that?

yasminvalim replied:
Hey Huijing, I removed this one in line 148. Should remove all _#[cfg(target_os = "linux")]_?

HuijingHei replied:
Yes, I think so.

HuijingHei commented:
There are a lot of changes since https://github.com/coreos/bootupd/pull/1068 merged, let us get CI and test [PR](https://github.com/coreos/bootupd/pull/1003) passed firstly. After that, would ask for help from Colin or Timothee to have a final review, WDYT?

yasminvalim committed:
Add copy-to-boot feature and add copy-to-boot cli command

This is part of Fedora BootLoaderUpdatesPhase1: https://fedoraproject.org/wiki/Changes/BootLoaderUpdatesPhase1

yasminvalim committed:
Check ESP space to avoid errors

yasminvalim committed:
Implement copy_efi_components_to_esp and add unit tests

yasminvalim committed:
docs: clarify package_mode_copy_to_boot; better statvfs errors. Use anyhow context for statvfs failures instead of raw_os_error.

AI-assisted: implementation reviewed by Claude. Some suggestions was added.

cgwalters commented on src/efi.rs:
```diff
@@ -87,9 +91,40 @@ pub(crate) fn is_efi_booted() -> Result<bool> {
         .map_err(Into::into)
 }
 
+fn mount_esp(esp_device: &Path, target: &Path) -> Result<()> {
```
This is basically just the `mount` system call as written I think, and we already have ESP mounting code, it's not clear to me why we need a new mounting functionality here?

That said, one thing that we *could* do here by using the new mount api (instead of raw `mount`) is to skip the `move_mount` part, and then just access everything via `mount_fd` (which we wrap in a `cap-std::fs::Dir`) or so.

cgwalters commented on src/efi.rs:
```diff
@@ -180,12 +225,25 @@ impl Efi {
+                        .run_inherited()
+                        .with_context(|| format!("Failed to unmount {:?}", mount.path))?;
+                    log::trace!("Unmounted");
+                }
```
I think `umount` should work, if it doesn't, forking off the external binary isn't more likely to work. Can you motivate this change? I think we should stick with one or the other.

If we're hitting issues because it's still in use, one option is to do a lazy unmount. But if it is in use it's likely because we still hold a file descriptor open, which we should have closed.

cgwalters commented on src/util.rs:
```diff
@@ -51,9 +53,17 @@ pub(crate) fn filenames(dir: &openat::Dir) -> Result<HashSet<String>> {
+/// Return the available space in bytes on the filesystem containing the given directory.
+/// Uses f_bavail * f_frsize from fstatvfs to avoid partial updates when the partition is full.
+pub(crate) fn available_space_bytes(dir: &openat::Dir) -> Result<u64> {
+    let fd = unsafe { BorrowedFd::borrow_raw(dir.as_raw_fd()) };
```
I think this can just use `.as_fd()`

yasminvalim replied:
openat::Dir is only AsRawFd, not AsFd, so I added a wrapper that implements AsFd and used `fstatvfs(wrapper(dir).as_fd())`. Is that okay or should I avoid using a wrappper? Is that any other way that I could implement it?

cgwalters commented on src/efi.rs:
```diff
@@ -236,6 +294,119 @@ impl Efi {
+    ) -> Result<()> {
+        // Build a merged source tree in a temp dir (same layout as desired ESP/EFI)
+        let temp_dir = tempfile::tempdir().context("Creating temp dir for EFI merge")?;
+        let temp_efi_path = temp_dir.path().join("EFI");
```
It's a bit unfortunate copy all the files into a tempdir in a different place, then copy them again.

cgwalters commented on src/efi.rs:
```diff
@@ -236,6 +294,119 @@ impl Efi {
+                efi_comp.version
+            );
+            // Copy contents of component's EFI dir (e.g. fedora/) into temp_efi_path so merged
+            // layout is EFI/fedora/..., not EFI/EFI/fedora/...
```
But, we could adjust how we do the copy so it's the right path to start.

cgwalters commented on src/efi.rs:
```diff
@@ -236,6 +294,119 @@ impl Efi {
+        let mut diff = current_filetree
+            .diff(&source_filetree)
+            .context("Computing EFI diff")?;
+        diff.removals.clear();
```
Why is this? Won't this mean we keep leaking old files?

cgwalters commented on src/efi.rs:
```diff
@@ -686,6 +853,11 @@ impl Component for Efi {
+    /// Package mode: merge `/usr/lib/efi` onto the ESP (write alongside, then atomic rename).
+    fn package_mode_copy_to_boot(&self, root: &Path) -> Result<()> {
+        self.package_mode_copy_to_boot_impl(root)
+    }
```
Why the indirection here?

yasminvalim committed:
Add reviewers suggestions

yasminvalim committed:
mend

yasminvalim committed:
mend
